src/Talis/Persona/Client/Tokens.php

```diff
@@ -291,6 +306,39 @@ public function isPresignedUrlValid($url, $secret)
         return $valid;
     }
 
+    /**
+     * List all scopes that belong to a given token
+     * @param string $token JWT token
+     * @oaram int $pubCertCacheTTL optional JWT public certificate time to live
+     * @return array list of scopes
+     *
+     * @throws InvalidValidationException invalid signature, key or token
+     * @throws \DomainException decoded token or metadata does not adhere to
+     * domain models
+     */
+    public function listScopes($token, $pubCertCacheTTL = 300)
+    {
+        $publicCert = $this->retrieveJWTCertificate($pubCertCacheTTL);
+        $encodedToken = $token['access_token'];
+        $decodedToken = $this->decodeToken($encodedToken, $publicCert);
+
+        if (isset($decodedToken['scopes']) && is_array($decodedToken['scopes'])) {
+            return $decodedToken['scopes'];
+        }
+
+        if (isset($decodedToken['scopeCount'])) {
+            $meta = $this->personaRetrieveTokenMetadata($encodedToken);
+
+            if (isset($meta['scopes']) && is_string($meta['scopes'])) {
+                return explode(' ', $meta['scopes']);
+            }
+
+            throw new \DomainException('token metadata missing scopes attribute');
+        }
+
+        throw new \DomainException('decoded token is both scope attributes');
```

Comment:
"... is *missing* both scope attributes"?

(If that's the intended meaning, how about "decoded token has neither `scopes` nor `scopeCount`".)

I don't know if `\DomainException` is right here - it's failed validation, so I'd expect it to throw an `InvalidValidationException` - one exception type to rule them all.  Do we somehow need `DomainException` to distinguish internally?

Reply:
(This feels like a `InvalidTokenException`, by my understanding of your new code)

Reply:
@junglebarry I would like to use InvalidTokenException, but it would cause a breaking change, which I would not like to include in this issue as it is off topic.

Reply:
Ignore me ... I can change it.